src/Compilers/CSharp/Portable/BoundTree/NullabilityRewriter.cs

```diff
@@ -8,5 +13,68 @@ protected override BoundExpression VisitExpressionWithoutStackGuard(BoundExpress
         {
             return (BoundExpression)Visit(node);
         }
+
+        public override BoundNode VisitBinaryOperator(BoundBinaryOperator node)
+        {
+            return VisitBinaryOperatorBase(node, (binary, left, right, type) => node.Update(node.OperatorKind, node.ConstantValueOpt, node.MethodOpt, node.ResultKind, left, right, type));
+        }
+
+        public override BoundNode VisitUserDefinedConditionalLogicalOperator(BoundUserDefinedConditionalLogicalOperator node)
+        {
+            return VisitBinaryOperatorBase(node, (binary, left, right, type) => node.Update(node.OperatorKind, node.LogicalOperator, node.TrueOperator, node.FalseOperator, node.ResultKind, left, right, type));
+        }
+
+        private BoundNode VisitBinaryOperatorBase(BoundBinaryOperatorBase binaryOperator, Func<BoundBinaryOperatorBase, BoundExpression, BoundExpression, TypeSymbol, BoundBinaryOperatorBase> nodeUpdater)
+        {
+            // Use an explicit stack to avoid blowing the managed stack when visiting deeply-recursive
+            // binary nodes
+            var stack = ArrayBuilder<BoundBinaryOperatorBase>.GetInstance();
+            BoundBinaryOperatorBase currentBinary;
+            BoundExpression child = binaryOperator;
+
+            while (true)
+            {
+                if (child is BoundBinaryOperatorBase childBinary)
```

Comment:
>child is BoundBinaryOperatorBase childBinary [](start = 20, length = 44)

It looks like this condition can simply be moved into the ```while``` condition and the ```if``` can be removed. Also, it looks like we are always doing at least one iteration, so the first condition check is redundant and it can be avoided by using a loop with post condition. Consider the following refactoring:
```
BoundBinaryOperatorBase currentBinary = binaryOperator;
// Remove            BoundExpression child = binaryOperator;

do
{
    stack.Push(currentBinary);
    currentBinary= currentBinary.Left as BoundBinaryOperatorBase;
}
while (currentBinary != null);
```